sql/catalyst/src/main/java/org/apache/spark/sql/connector/write/WriteBuilder.java

```diff
@@ -23,17 +23,34 @@
 import org.apache.spark.sql.connector.write.streaming.StreamingWrite;
 
 /**
- * An interface for building the {@link BatchWrite}. Implementations can mix in some interfaces to
+ * An interface for building the {@link Write}. Implementations can mix in some interfaces to
  * support different ways to write data to data sources.
  *
- * Unless modified by a mixin interface, the {@link BatchWrite} configured by this builder is to
+ * Unless modified by a mixin interface, the {@link Write} configured by this builder is to
  * append data without affecting existing data.
  *
  * @since 3.0.0
  */
 @Evolving
 public interface WriteBuilder {
 
+  /**
+   * Returns a logical {@link Write} shared between batch and streaming.
+   */
+  default Write build() {
```

Comment:
The `buildForBatch` method (and stream equivalent) are already released, so this generic `Write` implementation makes the new structure, `build` + `toBatch`, work for existing sources. It also allows sources to implement the version that they choose. So if none of the features that require the `Write` are used, I guess they could avoid a mostly-boilerplate class.

Reply:
Correct, this method was introduced to keep the compatibility.

Reply:
Have you considered to change the default impl for `buildForBatch` to:
```scala
  default BatchWrite buildForBatch() {
    build().toBatch()
  }
```

and also the `build()` to just return a simple anonymous `new Write() {}`? 

Otherwise, I can see that we'll have the `buildForBatch` (and similarly `buildForStreaming`) logic in two different places: `WriteBuilder` and `Write`. It is easy to miss one or another.

Reply:
I am not sure I understood. Could you elaborate a bit more, @sunchao?

Spark will now always call `build()` and work with the `Write` abstraction. I added the default implementation so that  existing data sources that already implement the current API will continue to work as before. Spark will is not supposed to call `buildForBatch` after this change.

Reply:
New data sources should be encouraged to implement only `build`.

Reply:
We should probably deprecate the other ones.

Reply:
What I mean is now we can potentially have two copies of the `toBatch` implementation: one in `WriteBuilder.buildForBatch` and one in `Write.toBatch`, when users start to override `build`, `buildForBatch` and `buildForStreaming`. If moving forward we want `build` to be the canonical impl, perhaps we can make `buildForBatch` and `buildForStreaming` to just call `build.toBatch()` internally so that users just need to override `build`.